### PR TITLE
Add an assertion that verifies the size of cached debug info types.

### PR DESCRIPTION
**Pull request by adrian-prantl**
DebugTypeInfo: Prefer the size of the Storage type derived from the TypeInfo.
    
Previously type sizes would be inconsistently sourced from either the LLVM type
or the FixedTypeInfo, depending on the call site. This was problematic because
TypeInfo operates with a resolution of whole bytes, which means that types such
as i1 would get a reported as having a size of 8. This patch now asserts that
all occurrences of the same type have the same size as the first, cached
occurence.

To avoid triggering the cached type verification assertion, this patch avoids
caching of storage-sized containers. It also removes the unique identifier from
forward declarations, which could lead to type confusion during LTO.

rdar://102367872

### CODE REVIEW

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test source compatibility

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test source compatibility

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test source compatibility

**Comment by adrian-prantl** [comment]
@swift-ci test

File: lib/IRGen/IRGenDebugInfo.cpp [hunk 1]
```diff
@@ -138,7 +138,9 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
   /// \}
 
   /// A list of replaceable fwddecls that need to be RAUWed at the end.
-  std::vector<std::pair<TypeBase *, llvm::TrackingMDRef>> ReplaceMap;
+  std::vector<std::pair<TypeBase *, llvm::TrackingMDRef>> FwdDeclScopes;
```
Comment (augusto2112) [inline lib/IRGen/IRGenDebugInfo.cpp:141]: It doesn't look like anything is added in `FwdDeclScopes`?

**Review by augusto2112 — APPROVED** [review]

File: lib/IRGen/IRGenDebugInfo.cpp (hunk 1 above)
Comment (adrian-prantl) [reply lib/IRGen/IRGenDebugInfo.cpp:141]: Thanks! That's a leftover from an earlier version of the patch.

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test source compatibility

**Comment by adrian-prantl** [comment]
Nice, the only failure is an unrelated linker crash.

**Comment by adrian-prantl** [comment]
@swift-ci test and merge

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test macos

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test platform macOS

**Comment by adrian-prantl** [comment]
@swift-ci test macOS

**Comment by adrian-prantl** [comment]
@swift-ci test macOS

**Comment by adrian-prantl** [comment]
@swift-ci test and merge

**Pull request merged**